cmd/crypto/vault.go

```diff
@@ -130,70 +130,79 @@ func NewVault(config VaultConfig) (KMS, error) {
 	return v, nil
 }
 
-// reauthenticate() tries to login in 1 minute
-// intervals until successful.
-func (v *vaultService) reauthenticate() {
-	retryDelay := 1 * time.Minute
-	go func() {
-		for {
-			if err := v.authenticate(); err != nil {
-				time.Sleep(retryDelay)
-				continue
-			}
-			return
-		}
-	}()
-}
-
-// renewer calls vault client's renewer that automatically
-// renews secret periodically
-func (v *vaultService) renewer(secret *vault.Secret) {
+// renewSecret tries to renew the given secret. It blocks
+// until it receives either the new secret or encounters an error.
+func (v *vaultService) renewSecret(secret *vault.Secret) (*vault.Secret, error) {
 	renewer, err := v.client.NewRenewer(&vault.RenewerInput{
 		Secret: secret,
 	})
 	if err != nil {
-		logger.FatalIf(err, "crypto: hashicorp vault token renewer could not be started")
+		logger.CriticalIf(context.Background(), fmt.Errorf("crypto: failed to create hashicorp vault renewer: %s", err))
 	}
-	v.tokenRenewer = renewer
 	go renewer.Renew()
 	defer renewer.Stop()
 
 	for {
 		select {
 		case err := <-renewer.DoneCh():
 			if err != nil {
-				v.reauthenticate()
-				renewer.Stop()
-				return
+				return nil, err
 			}
-
-			// Renewal is now over
-		case renewal := <-renewer.RenewCh():
-			v.leaseDuration = time.Duration(renewal.Secret.Auth.LeaseDuration)
+		case renew := <-renewer.RenewCh():
+			if renew.Secret == nil || renew.Secret.Auth == nil {
+				return nil, ErrKMSAuthLogin
+			}
+			return renew.Secret, nil
 		}
 	}
 }
 
-// authenticate logs the app to vault, and starts the auto renewer
-// before secret expires
-func (v *vaultService) authenticate() (err error) {
+// login tries to authenticate the minio server to
+// the Vault KMS using the approle ID and secret.
+func (v *vaultService) login() (*vault.Secret, error) {
 	payload := map[string]interface{}{
 		"role_id":   v.config.Auth.AppRole.ID,
 		"secret_id": v.config.Auth.AppRole.Secret,
 	}
-	var secret *vault.Secret
-	secret, err = v.client.Logical().Write("auth/approle/login", payload)
+	secret, err := v.client.Logical().Write("auth/approle/login", payload)
 	if err != nil {
-		return
+		return nil, err
+	}
+	if secret == nil || secret.Auth == nil {
+		return nil, ErrKMSAuthLogin
 	}
-	if secret.Auth == nil {
-		err = ErrKMSAuthLogin
-		return
+	return secret, nil
+}
+
+// authenticate tries to authenticate the minio server
+// to the Vault KMS and starts a background job to renew
+// the login.
+func (v *vaultService) authenticate() error {
+	secret, err := v.login()
+	if err != nil {
+		return err
 	}
 	v.client.SetToken(secret.Auth.ClientToken)
 	v.leaseDuration = time.Duration(secret.Auth.LeaseDuration)
-	go v.renewer(secret)
-	return
+
+	// Start background job trying to renew the token
+	// or (if this fails) try to login again with app-ID and app-Secret.
+	go func(secret *vault.Secret) {
+		var err error
+		for {
+			secret, err = v.renewSecret(secret) // try to renew the secret (blocking)
```

Comment:
@aead, If v.Login() returned nil for secret, we would do v.renewSecret with secret as nil. This will stop minio server with
 ```ERROR crypto: hashicorp vault token renewer could not be started: missing secret to renew.```  because of the logger.CriticalIf in renewSecret.

Reply:
Isn't the assumption that login will never return `nil` for secret? - given that there will always be `err` defined?

Reply:
But `v.login()` never returns `secret == nil` and `err == nil`. In case of error we retry - otherwise `secret != nil`. Or am I missing something?! See: https://github.com/minio/minio/pull/7241/files/edb764e3a2f1ba27de6491d7c031ee886c2e3b0c#diff-bc9558a54da677c8f7933fee0f0c42f5R171

Reply:
Also criticalIf is caught it's not os.Exit(1), server will continue to operate for further API operations.

Reply:
I see it happening during testing - for any login error we return nil as secret. The go routine for loop uses this to renewSecret()

Reply:
@aead the possibility @poornas is talking about is when renewSecret() fails and also `v.login()` fails then we erroneously update a global `secret` value with `nil` - the value which was passed in the go-routine. 

Even in this scenario a panic with CriticalIf() just rejects the handler and not crash the server. We cannot avoid vault not responding to what we want. We can only do is crash the handler instead not the server which is what CriticalIf will do.

Reply:
As it stands though retry will not work in a fool proof manner here since `v.login()` can fail and then keep crashing the handlers.. 

so a better approach is to change the behavior and instead re-implement the go routine


```go
// authenticate tries to authenticate the minio server
// to the Vault KMS and starts a background job to renew
// the login.
func (v *vaultService) authenticate() error {
	secret, err := v.login()
	if err != nil {
		return err
	}
	v.client.SetToken(secret.Auth.ClientToken)
	v.leaseDuration = time.Duration(secret.Auth.LeaseDuration)

	// Start background job trying to renew the token
	// or (if this fails) try to login again with app-ID and app-Secret.
	go func(secret *vault.Secret) {
		var err error
		for {
			newSecret, err = v.renewSecret(secret) // try to renew the secret (blocking)
			if err != nil {
				// Try to login again with app-ID and app-Secret
				if newSecret, err = v.login(); err != nil { // failed -> try again
					time.Sleep(1 * time.Minute) // retry delay
					continue
				}
			}
			v.client.SetToken(newSecret.Auth.ClientToken)
			v.leaseDuration = time.Duration(newSecret.Auth.LeaseDuration)
                        secret = newSecret
		}
	}(secret)
	return nil
}
```

Reply:
Good catch - should be fixed now :+1: